Tempura/Navigation/NavigationWitness.swift

```diff
@@ -0,0 +1,182 @@
+//
+//  NavigationWitness.swift
+//  Tempura
+//
+//  Created by Daniele Formichelli on 18/11/20.
+//  Copyright © 2020 Bending Spoons. All rights reserved.
+//
+
+import Hydra
+import Katana
+
+// MARK: - Protocol Witness
+
+/// Tempura navigation protocol witness.
+public struct NavigationWitness {
+  private let show: ([RouteElementIdentifier], Bool, Any?) -> Promise<Void>
+  private let hide: (RouteElementIdentifier, Bool, Any?, Bool) -> Promise<Void>
+
+  /// @see Tempura.Show
+  public func show(
+    _ identifiersToShow: [RouteElementIdentifier],
+    animated: Bool = false,
+    context: Any? = nil
+  ) -> Promise<Void> {
+    return self.show(identifiersToShow, animated, context)
+  }
+
+  /// @see Tempura.Show
+  public func show(
+    _ identifiersToShow: RouteElementIdentifier,
+    animated: Bool = false,
+    context: Any? = nil
+  ) -> Promise<Void> {
+    return self.show([identifiersToShow], animated, context)
+  }
+
+  /// @see Tempura.Show
+  public func show<I>(
+    _ identifiersToShow: [I],
+    animated: Bool = false,
+    context: Any? = nil
+  ) -> Promise<Void>
+    where I: RawRepresentable, I.RawValue == RouteElementIdentifier {
+    return self.show(identifiersToShow.map { $0.rawValue }, animated, context)
+  }
+
+  /// @see Tempura.Show
+  public func show<I>(
+    _ identifiersToShow: I,
+    animated: Bool = false,
+    context: Any? = nil
+  ) -> Promise<Void>
+    where I: RawRepresentable, I.RawValue == RouteElementIdentifier {
+    return self.show([identifiersToShow.rawValue], animated, context)
+  }
+
+  /// @see Tempura.Hide
+  public func hide(
+    _ identifierToHide: RouteElementIdentifier,
+    animated: Bool = false,
+    context: Any? = nil,
+    atomic: Bool = false
+  ) -> Promise<Void> {
+    return self.hide(identifierToHide, animated, context, atomic)
+  }
+
+  /// @see Tempura.Hide
+  public func hide<I>(
+    _ identifierToHide: I,
+    animated: Bool = false,
+    context: Any? = nil
+  ) -> Promise<Void>
+    where I: RawRepresentable, I.RawValue == RouteElementIdentifier {
+    return self.hide(identifierToHide.rawValue, animated, context, false)
+  }
+
+  /// @see Tempura.Hide
+  public func hide(
+    animated: Bool = false,
+    context: Any? = nil,
+    atomic: Bool = false
+  ) -> Promise<Void> {
+    guard let identifierToHide = UIApplication.shared.currentRoutableIdentifiers.last else {
+      fatalError("No routable identifiers found")
+    }
+    return self.hide(identifierToHide, animated, context, atomic)
+  }
+}
+
+// MARK: - Live
+
+extension NavigationWitness {
```

Comment:
why are we not wrapping this in a `#if DEBUG`? this stuff should NEVER reach prod

Reply:
The live one needs to, actually. But I agree about the `mocked`, the `spy` and the `unimplemented`